_pytest/cacheprovider.py

```diff
@@ -168,6 +170,41 @@ def pytest_sessionfinish(self, session):
             config.cache.set("cache/lastfailed", self.lastfailed)
 
 
+class NFPlugin(object):
+    """ Plugin which implements the --nf (run new-first) option """
+
+    def __init__(self, config):
+        self.config = config
+        self.active = config.option.newfirst
+        self.all_items = config.cache.get("cache/allitems", {})
+
+    def pytest_collection_modifyitems(self, session, config, items):
+        if self.active:
+            new_items = []
+            other_items = []
+            for item in items:
+                mod_timestamp = os.path.getmtime(str(item.fspath))
+                if self.all_items and item.nodeid not in self.all_items:
+                    new_items.append((item, mod_timestamp))
+                else:
+                    other_items.append((item, mod_timestamp))
+
+            items[:] = self._get_increasing_order(new_items) + \
+                self._get_increasing_order(other_items)
+        self.all_items = items
+
+    def _get_increasing_order(self, test_list):
+        test_list = sorted(test_list, key=lambda x: x[1], reverse=True)
+        return [test[0] for test in test_list]
+
+    def pytest_sessionfinish(self, session):
+        config = self.config
+        if config.getoption("cacheshow") or hasattr(config, "slaveinput"):
+            return
+        config.cache.set("cache/allitems",
+                         [item.nodeid for item in self.all_items if isinstance(item, Function)])
```

Comment:
Why are you testing `isinstance` here? 🤔

Reply:
Please, check testing/acceptance_test.py:584
https://github.com/pytest-dev/pytest/blob/063e2da967a715b06b912586ba7eb50464984ac7/testing/acceptance_test.py#L580-L585
It failes because sometimes item is a string, not object:
```
AttributeError: 'unicode' object has no attribute 'nodeid'
```

Reply:
I see thanks. Sometimes `pytest_collect_modifyitems` can received strings, weird. Didn't take the time to investigate this further, but regardless I think it is safer to check for `pytest.Item` instead while still in `pytest_collect_modifyitems`:

```python
self.cached_nodeids = [x.nodeid for x in items if isinstance(x, pytest.Item)]
```

(Also already suggesting renaming `all_items` to `cached_nodeids`). During `pytest_sessionfinish` you can dump it directly:

```python
config.cache.set("cache/allitems", self.cached_nodeids )
```